### PR TITLE
Fix Integrations Bugs

### PR DESCRIPTION
**Pull request by georgreen**
## Resolves #<issue number>

Fixes [#158909836 #158909836  #158909836 #159006992]

## Description (what problem you're fixing)

  - #158790636 - Fix roles shape: Incompatible payload between front-end and backend
  - #158790655 - Fix Auth: user should only be saved once: Bg: users were saved mutiple times leading to losing their states in the DB as this is effectivly wiping the user then save them in DB
  - #158909836 - Update auth to use new token structure: Andela tech team update their api which broke the app
  - Add migrations to app: Previously tracking changes to models was not being tracked
  - Update seed data command: Previously the command wipped your DB before it dumped data

## Fix (what you did to fix it)

  - #158790636 : sync with the front-end shape
  - #158790655: fixed the logic, add a check to stop mutiple saves
 - #158909836: provide the required fields for compatibility with the tech team api 
- Refactored code base: fixed failing tests and broken invariants after changes 

## How to test (describe how to test your PR)

- Clone the repository, `install requirements` within the repository.
- `$ python manage.py tests`

### CODE REVIEW

File: src/api/endpoints/activity_types.py [hunk 1]
```diff
@@ -35,11 +35,13 @@ def post(cls):
             validation_status_code = status_code or 400
             return response_builder(errors, validation_status_code)
 
-        activity_type = ActivityType(name=result["name"],
-                                     description=result["description"],
-                                     value=result["value"],
-                                     supports_multiple_participants=result["supports_multiple_participants"]
-                                     )
+        supprt_mutiple = result["supports_multiple_participants"]
```
Comment (Andretalik) [inline src/api/endpoints/activity_types.py]: Typo at `supprt` meaning there is a typo copied over and over again.

File: src/api/endpoints/activity_types.py [hunk 2]
```diff
@@ -35,11 +35,13 @@ def post(cls):
...
+        activity_type = ActivityType(
+                name=result["name"],
+                description=result["description"],
+                value=result["value"],
+                supports_multiple_participants=supprt_mutiple
```
Comment (Andretalik) [inline src/api/endpoints/activity_types.py]: Here is the other typo.

File: src/api/endpoints/redemption_requests.py [hunk 3]
```diff
@@ -27,9 +45,9 @@ def post(cls):
                 status="fail"
             ), 400)
 
-        name = payload.get("name")
+        name = payload.get("reason")
         value = payload.get("value")
-        country_input = payload.get("country")
+        country_input = payload.get("center")
 
         country = Country.query.filter_by(name=country_input).first()
```
Comment (Andretalik) [inline src/api/endpoints/redemption_requests.py]: Shouldn't this be referencing Center? The work on Countries being renamed was merged.

File: src/api/endpoints/redemption_requests.py [hunk 4]
```diff
@@ -38,16 +56,18 @@ def post(cls):
                 name=name,
                 value=value,
                 user=g.current_user,
-                country=country
+                country=country,
```
Comment (Andretalik) [inline src/api/endpoints/redemption_requests.py]: Center, here too

File: src/api/endpoints/redemption_requests.py [hunk 5]
```diff
@@ -38,16 +56,18 @@ def post(cls):
...
+                society=g.current_user.society
```
Comment (Andretalik) [inline src/api/endpoints/redemption_requests.py]: Is it necessary to add this when we can query the Society from the user?
Why add more relationships?

File: src/api/endpoints/redemption_requests.py [hunk 6]
```diff
@@ -103,60 +123,56 @@ def put(cls, redeem_id=None):
...
-            search_term_name = request.args.get('name')
+            search_term_name = request.args.get('society')
             if search_term_name:
-                redemp_request = RedemptionRequest.query.filter_by(
+                society = Society.query.filter_by(
                                         name=search_term_name).first()
-                return find_item(redemp_request)
+                if not society:
+                    mes = f"Society with name:{search_term_name} not found"
+                    return {"message": mes}, 400
```
Comment (Andretalik) [inline src/api/endpoints/redemption_requests.py]: We should standardise how we build our responses.
Wasn't this the function of the response builder?

File: src/api/endpoints/redemption_requests.py [hunk 7]
```diff
@@ -103,60 +123,56 @@ def put(cls, redeem_id=None):
...
+                                        name=search_term_name)
+                return redemption_paginate(redemp_request)
 
             search_term_country = request.args.get("country")
             if search_term_country:
                 country_query = Country.query.filter_by(
                             name=search_term_country).first()
+                if not country_query:
+                    mes = f"country with name:{search_term_name} not found"
+                    return {"message": mes}, 400
```
Comment (Andretalik) [inline src/api/endpoints/redemption_requests.py]: Still on standardising the responses being built

File: src/api/endpoints/users.py [hunk 8]
```diff
@@ -63,5 +63,9 @@ def get(cls, user_id=None):
         user_information['location'] = location
         user_information['cohort'] = cohort_serialized
         user_information['society'], _ = basic_info_schema.dump(cohort.society)
+        # import pdb; pdb.set_trace()
```
Comment (Andretalik) [inline src/api/endpoints/users.py]: Commented out code shouldn't make it here.

File: src/api/endpoints/users.py [hunk 9]
```diff
@@ -63,5 +63,9 @@ def get(cls, user_id=None):
...
+
+        user_information['roles'] = {role['name']: role['id']
+                                     for role in user_information['roles']}
 
         return dict(data=user_information), status_code
```
Comment (Andretalik) [inline src/api/endpoints/users.py]: Still on standardsing responses.

File: src/api/models.py [hunk 10]
```diff
@@ -175,6 +175,8 @@ class Society(Base):
     logged_activities = db.relationship('LoggedActivity', backref='society',
                                         lazy='dynamic')
     cohorts = db.relationship('Cohort', backref='society', lazy='dynamic')
+    redemptions = db.relationship('RedemptionRequest', backref='society',
```
Comment (Andretalik) [inline src/api/models.py]: Still on the relationship question asked earlier.

File: src/api/utils/auth.py [hunk 11]
```diff
@@ -115,24 +121,25 @@ def store_user_details(payload, token):
...
+            route_roles = list(filter(lambda r: r is not None,
+                                      [Role.query.filter_by(name=role).first()
+                                       for role in roles]))
+            for role in g.current_user.roles:
+                if role in route_roles:
+                    break
+            else:
+                return response_builder(
+                    dict(message="You're unauthorized"
+                         " to perform this operation"), 401)
```
Comment (Andretalik) [inline src/api/utils/auth.py]: Could you walk me thorough how this works?

File: src/api/utils/auth.py [hunk 12]
```diff
@@ -48,9 +49,10 @@ def decorated(*args, **kwargs):
                                  options={
                                      'verify_signature': True,
                                      'verify_exp': True
-                                 }
+                                 },
+                                 audience=current_app.config['API_IDENTIFIER'],
+                                 issuer=current_app.config['API_IDENTIFIER']
```
Comment (Andretalik) [inline src/api/utils/auth.py]: Could you walk me through this as well?

File: src/tests/test_point_redemption.py [hunk 13]
```diff
@@ -179,24 +179,24 @@ def test_get_non_existing_redemption_requests_by_name(self):
...
                     content_type='application/json')
 
-        message = "does not exist"
+        message = "fetched successfully"
         response_details = json.loads(response.data)
 
         self.assertIn(message, response_details["message"])
-        self.assertEqual(response.status_code, 404)
+        self.assertEqual(response.status_code, 200)
+        self.assertFalse(response_details['data'])
```
Comment (Andretalik) [inline src/tests/test_point_redemption.py]: Why?

File: src/api/utils/auth.py (hunk 11 above)
Comment (georgreen) [reply src/api/utils/auth.py]: @Andretalik Loops through required roles and user roles, if the user has any of the required roles the loop break and user is allowed to carry out the operation if not and the loop runs through the else block executes.

File: src/api/endpoints/activity_types.py (hunk 1 above)
Comment (georgreen) [reply src/api/endpoints/activity_types.py]: @Andretalik Will be fixed

File: src/api/utils/auth.py (hunk 12 above)
Comment (georgreen) [reply src/api/utils/auth.py]: @Andretalik Tech team changed their token structure to include that but those changes have been reverted.

File: src/tests/test_point_redemption.py (hunk 13 above)
Comment (georgreen) [reply src/tests/test_point_redemption.py]: This is a filter, 404 suggest the `url` is malformed which is not the case, I think it means the user passed the wrong data in the `url`.

**Comment by kimani-ndegwa** [comment]
Trailing whitespace? @georgreen

File: src/prod-stag-migrations/versions/674aca47c075_.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
+"""empty message
+
+Revision ID: 674aca47c075
+Revises:
```
[the remark this answers is not in the document]
Comment (georgreen) [reply src/prod-stag-migrations/versions/674aca47c075_.py]: These are migration files, they are auto-generated, we will remove migration files from coday analysis, for now, I don't have the permissions to do that but I'm working on it. @kimani-ndegwa

File: src/api/endpoints/activity_types.py [hunk 15]
```diff
@@ -1,13 +1,13 @@
 """Module for activity types operations."""
 
 from flask import request
-from flask_restplus import Resource
```
Comment (LarryWachira) [inline src/api/endpoints/activity_types.py]: Didn't know you'd moved from restplus. Any reason?

File: src/api/utils/helpers.py [hunk 16]
```diff
@@ -87,23 +88,21 @@ def paginate_items(fetched_data):
...
         return response_builder(dict(
             status="success",
-            data=dict(data_items=data_list,
-                      count=len(fetched_data.items),
-                      nextUrl=next_url,
-                      previousUrl=previous_url,
-                      currentPage=fetched_data.page),
-            message="Data fetched successfully."
+            data=data_list,
+            count=len(fetched_data.items),
```
Comment (LarryWachira) [inline src/api/utils/helpers.py]: Sqlalchemy's pagination object has a `total` attribute. Also `pages` if you'd want to know how many pages of the results there are.

File: src/manage.py [hunk 17]
```diff
@@ -53,18 +54,38 @@ def seed():
         "Testing": test_data,
         "Staging": production_data
     }
-    if environment == "Testing" or \
-        prompt_bool("\n\n\nThis operation will remove all existing data."
-                    " Are you sure you want to continue?"):
+
+    if environment.lower() in ["production", "staging"]:
```
Comment (LarryWachira) [inline src/manage.py]: Are there plans to remove this once production is up? I have mad jitters about leaving it in.

File: src/api/endpoints/activity_types.py (hunk 15 above)
Comment (georgreen) [reply src/api/endpoints/activity_types.py]: We had decided on this a while back, also we are using apiary for our documentation.

File: src/api/utils/helpers.py (hunk 16 above)
Comment (georgreen) [reply src/api/utils/helpers.py]: Can I add that in my next PR?

Comment (LarryWachira) [reply src/api/utils/helpers.py]: No problem. I'll remind you.

**Pull request merged**